### PR TITLE
[InstCombine] Fold `select (or/and (icmp eq/ne a, b), other_cond), a, b` to `select (other_cond), a, b`

### LINKED ISSUE #75784: [InstCombine] Missing optimization: fold `select (a != V1 & a != V2), V1, a` to `select (a == V2), V2, V1`
Alive2 proof: https://alive2.llvm.org/ce/z/4_WZ9s

### Description:

```llvm
define i32 @src(i32 %a, i32 noundef %v1, i32 noundef %v2)  {
entry:
  %cmp = icmp ne i32 %a, %v1
  %cmp1 = icmp ne i32 %a, %v2
  %or.cond = and i1 %cmp, %cmp1
  %select = select i1 %or.cond, i32 %v1, i32 %a
  ret i32 %select
}
```
can be folded to:
```llvm
define i32 @tgt(i32 %a, i32 noundef %v1, i32 noundef %v2)  {
entry:
  %switch.selectcmp = icmp eq i32 %a, %v2
  %switch.select = select i1 %switch.selectcmp, i32 %v2, i32 %v1
  ret i32 %switch.select
}
```

### Real-world motivation

This snippet of IR is derived from [linux/lib/hexdump.c@hex_dump_to_buffer](https://github.com/torvalds/linux/blob/ceb6a6f023fd3e8b07761ed900352ef574010bcb/lib/hexdump.c#L127) (after O3 pipeline).
The example above is a reduced version. If you're interested in the original suboptimal IR and optimal IR, see also:https://godbolt.org/z/8K3qqzMzr

**Let me know if you can confirm that it's an optimization opportunity, thanks.**

### PR DESCRIPTION
**Pull request by XChy**
Fixes #75784
Alive2 proof: https://alive2.llvm.org/ce/z/XduDgy

### CODE REVIEW

**Review by nikic — CHANGES_REQUESTED** [review]
This transform looks unnecessarily specific to me. I think what you actually want to do is this: https://alive2.llvm.org/ce/z/sN48Rm

We only need the one icmp in the condition, the other one is irrelevant. If the other icmp allows an operand replacement in the select, we already have a separate fold that will do that (select value equivalence fold).

**Comment by XChy** [comment]
> This transform looks unnecessarily specific to me. I think what you actually want to do is this: https://alive2.llvm.org/ce/z/sN48Rm
> 
> We only need the one icmp in the condition, the other one is irrelevant. If the other icmp allows an operand replacement in the select, we already have a separate fold that will do that (select value equivalence fold).

Sounds great to me! It also applies to `and(cmp, icmp ne)`, with [proof](https://alive2.llvm.org/ce/z/k6uRw-). I wonder how you generalize these patterns, just replace the component with argument randomly?

**Comment by XChy** [comment]
Complete alive2 proof for commutated version: https://alive2.llvm.org/ce/z/CVH7kK

File: llvm/test/Transforms/InstCombine/select-and-or.ll [hunk 1]
```diff
@@ -965,3 +965,116 @@ define i1 @or_and3_wrong_operand(i1 %a, i1 %b, i32 %x, i32 %y, i1 %d) {
...
+  %select = select i1 %cond, i32 %a, i32 %b
+  ret i32 %select
+}
+
+define i32 @and_ne_a_b(i1 %other_cond, i32 %a, i32 %b)  {
+; CHECK-LABEL: @and_ne_a_b(
+; CHECK-NEXT:    [[SELECT:%.*]] = select i1 [[OTHER_COND:%.*]], i32 [[A:%.*]], i32 [[B:%.*]]
+; CHECK-NEXT:    ret i32 [[SELECT]]
+;
+  %cmp = icmp eq i32 %a, %b
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/select-and-or.ll:985]: Was this supposed to be ne?

File: llvm/test/Transforms/InstCombine/select-and-or.ll [hunk 2]
```diff
@@ -965,3 +965,116 @@ define i1 @or_and3_wrong_operand(i1 %a, i1 %b, i32 %x, i32 %y, i1 %d) {
...
+  %cond = or i1 %other_cond, %cmp
+  %select = select i1 %cond, i32 %a, i32 %b
+  ret i32 %select
+}
+
+define i32 @or_eq_a_b_mutated(i1 %other_cond, i32 %a, i32 %b)  {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/select-and-or.ll:991]:
```suggestion
define i32 @or_eq_a_b_commuted(i1 %other_cond, i32 %a, i32 %b)  {
```

File: llvm/test/Transforms/InstCombine/select-and-or.ll [hunk 3]
```diff
@@ -965,3 +965,116 @@ define i1 @or_and3_wrong_operand(i1 %a, i1 %b, i32 %x, i32 %y, i1 %d) {
...
+; CHECK-NEXT:    [[SELECT:%.*]] = select i1 [[COND]], i32 [[B]], i32 [[A]]
+; CHECK-NEXT:    ret i32 [[SELECT]]
+;
+entry:
+  %cmp = icmp ne i32 %a, %c
+  %cmp1 = icmp ne i32 %b, %c
+  %cond = and i1 %cmp, %cmp1
+  %select = select i1 %cond, i32 %b, i32 %a
+  ret i32 %select
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/select-and-or.ll:1080]: Missing test coverage for and (apart from this negative test)?

**Review by nikic — COMMENTED** [review]
It should be possible to generalize the fold via simplifyWithOpReplaced(), but this seems like a reasonable starting point.

File: llvm/test/Transforms/InstCombine/select-and-or.ll (hunk 1 above)
Comment (XChy) [reply llvm/test/Transforms/InstCombine/select-and-or.ll]: Yep. It seems to change after rebase... Done now.

File: llvm/test/Transforms/InstCombine/select-and-or.ll [hunk 4]
```diff
@@ -965,3 +965,139 @@ define i1 @or_and3_wrong_operand(i1 %a, i1 %b, i32 %x, i32 %y, i1 %d) {
...
+
+define i32 @or_slt_a_b_fail(i1 %other_cond ,i32 %a, i32 %b)  {
+; CHECK-LABEL: @or_slt_a_b_fail(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[CMP:%.*]] = icmp slt i32 [[A:%.*]], [[B:%.*]]
+; CHECK-NEXT:    [[COND:%.*]] = or i1 [[CMP]], [[OTHER_COND:%.*]]
+; CHECK-NEXT:    [[SELECT:%.*]] = select i1 [[COND]], i32 [[A]], i32 [[B]]
+; CHECK-NEXT:    ret i32 [[SELECT]]
+; CHECK:       0:
```
Comment (XChy) [inline llvm/test/Transforms/InstCombine/select-and-or.ll:1065]: I don't know why there are extra checks, but utils/update_test_checks.py indeed generates it.

**Comment by XChy** [comment]
ping.

**Comment by dtcxzyw** [comment]
It also holds for logical and/or.
Alive2: https://alive2.llvm.org/ce/z/9Ko8wb

**Comment by dtcxzyw** [comment]
Emm, I seem to have found a better approach. Will post a patch later.

**Comment by XChy** [comment]
#76621 has fixed this issue.

**Pull request closed without merging**